### ISSUE TITLE
CourseAPI: Valid subjects/terms are labeled invalid if no courses exist

### ISSUE THREAD

**Issue by Rahi374** [Pitt-CSC/PittAPI#22]
I caught this while unit testing.
The specific example is subject="ARABIC" and term="2171" or term="2177". For these two terms, no Arabic courses are offered, so the array `course_details` is empty, and the code then assumes that because the array is empty, then there must be an invalid parameter and it throws an `InvalidParameteException`

**Comment by RitwikGupta** [comment]
As seen [here](http://www.courses.as.pitt.edu/results-subja.asp?TERM=2171&SUBJ=ARABIC), maybe we could look for "No records matched your query." and return an empty list?

**Comment by azharichenko** [comment]
The easiest way to prevent this is to have a list of all valid subjects and check the subject entered against it. The issue with @RitwikGupta solution is that `"No records matched your query."` always occurs when an invalid value is entered. As shown [by this](http://www.courses.as.pitt.edu/results-subja.asp?TERM=2171&SUBJ=A).

**Comment by azharichenko** [comment]
@Rahi374 I took a look at your unit test and it only address all correct scenarios, I'm writing up a slightly different version that includes your code but with fake subjects and term to make sure that it errors out properly.

**Comment by azharichenko** [comment]
Here is my current solution to the issue, `SUBJECT` is a list of all valid subjects.
```python
if not course_details:
    if subject in SUBJECTS:
        return []
    raise ValueError("The TERM or SUBJECT is invalid")
```

**Comment by Rahi374** [comment]
@azharichenko No I haven't yet pushed my unit tests that test this.

**Comment by Rahi374** [comment]
https://github.com/Rahi374/PittAPI/commit/40de5163f5ccba7e814a5a6063941b8dabd0c978

Here.
I guess I do also use the length of the course_list to test that retrieve_from_url though...

**Comment by Rahi374** [comment]
@azharichenko 

> but with fake subjects and term to make sure that it errors out properly.

Yeah that's probably a better idea. I hadn't gotten there yet.

> since your test only tests subject there are many failures do to using program code. I'm adding another part to the test where it will make sure that get_courses() with properly also get program courses.

What do you mean?

**Comment by azharichenko** [comment]
@Rahi374 
This is my method of testing each subject. 
```python
# Tests that all subjects properly return their course details
        for t in self.TERMS:
            for s in self.SUBJECTS:
                try:
                    results = course.get_courses(term=t, subject=s)
                except ValueError:
                    self.fail(msg="Term {} and/or Subject {} is not valid.".format(t, s))

                for result in results:
                    self.assertIn(t, result['term'], msg=s)
                    self.assertIn(s, result['subject'], msg=result)
```

When I started to do some initial testing I found that certain `subjects` you have defined are not really subject but actually programs which throw the test off. Some of these `programs` include `['BCC', 'CGSDAY', 'GSWS', 'CGSSAT']`. This is mainly because the classes in each `program` don't take the program code but instead, they use subject codes. 

So, for example, a `subject` like `BIOSC` has all its courses with the `BIOSC` subject code.
[As shown here](http://www.courses.as.pitt.edu/results-subja.asp?TERM=2171&SUBJ=BIOSC)

But for a `program` like `GSWS` it has courses with various codes such as `['ADMPS', 'AFRCNA', 'ANTH', 'ENGCMP', 'ENGLIT']`
[As shown here](http://www.courses.as.pitt.edu/results-subjspecial.asp?TERM=2171&SUBJ=GSWS)

Even further on a program page, it doesn't actually show all courses unless specified.
- results-subjspecial.asp?TERM=2171&SUBJ=GSWS : First 15 results
- results-subjspecia**la**.asp?TERM=2171&SUBJ=GSWS : All Results

---
So this is what I was talking about when I said the test might get a little more complicated because of this.

**Comment by RitwikGupta** [comment]
@azharichenko I targeted Python 2.7 when creating it, but over time we've ensured it fully supports Python 3 as well (check with `python3 -m py_compile pittAPI.py`). However, this puts us at a crossroads where we can't use things like `asyncio` and more to maintain the 2.7 support.

**Comment by RitwikGupta** [comment]
Closed with #43
